Fix HistoryController uses deprecated methods, #5004

low-batt opened the pull request:
This commit will:
- Change PlaybackHistory to conform to NSSecureCoding
- Change HistoryController to use methods that support secure coding instead of the deprecated methods

- [x] I have read [CONTRIBUTING.md](https://github.com/iina/iina/blob/develop/CONTRIBUTING.md)
- [x] This implements/fixes issue #5004.

---

**Description:**

low-batt committed:
Fix HistoryController uses deprecated methods, #5004

This commit will:
- Change PlaybackHistory to conform to NSSecureCoding
- Change HistoryController to use methods that support secure coding
  instead of the deprecated methods

Co-authored-by: Matt Svoboda <matt.svoboda@gmail.com>

uiryuu commented on iina/PlaybackHistory.swift:
```diff
@@ -15,7 +15,13 @@ fileprivate let KeyPlayed = "IINAPHPlayed"
+/// An entry in the playback history file.
+/// - Important: This class conforms to [NSSecureCoding](https://developer.apple.com/documentation/foundation/nssecurecoding).
+///     When making changes be certain the requirements for secure coding are not violated by the changes.
+class PlaybackHistory: NSObject, NSCoding, NSSecureCoding {
```
`NSSecureCoding` already inherits from `NSCoding`, do we need to confirm these two at the same time?

low-batt replied:
Sorry I missed that. Fixed.

low-batt committed:
Fix HistoryController uses deprecated methods, #5004

This commit will:
- Change PlaybackHistory to conform to NSSecureCoding
- Change HistoryController to use methods that support secure coding
  instead of the deprecated methods

Co-authored-by: Matt Svoboda <matt.svoboda@gmail.com>

uiryuu commented on iina/HistoryController.swift:
```diff
@@ -24,15 +24,26 @@ class HistoryController: NSObject {
+      let data = try Data(contentsOf: plistURL)
+      let object = try NSKeyedUnarchiver.unarchivedObject(ofClasses: [NSArray.self, PlaybackHistory.self],
+                                                          from: data)
+      history = object as? [PlaybackHistory] ?? []
```
I think we should throw an error there if `as?` fails so that we can log the type casting failed, tho I don't think that will happen very often...

uiryuu reviewed (approved):
You can decide whether or not to throw an error, I'm okay with both.

svobs commented:
Pulled this PR and ran it, added an item to the play history so that it was saved in the secure-encoded format, then quit & opened IINA 1.3.5 & IINA 1.3.4. Looks like older versions can still read it seamlessly. Nice.

low-batt commented on iina/HistoryController.swift:
```diff
@@ -24,15 +24,26 @@ class HistoryController: NSObject {
+    guard FileManager.default.fileExists(atPath: plistURL.path) else { return }
+    do {
+      let data = try Data(contentsOf: plistURL)
+      let object = try NSKeyedUnarchiver.unarchivedObject(ofClasses: [NSArray.self, PlaybackHistory.self],
+                                                          from: data)
+      history = object as? [PlaybackHistory] ?? []
```
I coded it that way with error logging and then simplified it and put it back to the way the code originally was. I prefer simplifying further to the code below. The other error handling needs to be there as you can never assume an external file hasn't been messed with. But the call to `unarchivedObject` will throw an error if the file contains any classes other than an array or `PlaybackHistory` objects. A developer would have to change `save` to save something like nested arrays, or just a single  `PlaybackHistory` object to cause `unarchivedObject` to be successful and the cast to fail. I don't see the need to cover that case, but I'm happy to add in the check and logging if you prefer an more cautious approach. Thoughts?
```suggestion
      history = try NSKeyedUnarchiver.unarchivedObject(ofClasses: [NSArray.self, PlaybackHistory.self],
                                                       from: data) as! [PlaybackHistory]
```

lhc70000 replied:
I think we don't need extra handling here.

uiryuu commented:
Thanks!
